plugins/elantp/fu-elantp-hid-haptic-device.c

```diff
@@ -0,0 +1,1037 @@
+
+	if (!fu_elantp_hid_haptic_device_write_cmd(parent,
+						   ETP_CMD_I2C_EEPROM_SETTING,
+						   ETP_CMD_I2C_EEPROM_WRITE_INFOMATION,
+						   error)) {
+		g_prefix_error(error, "failed to write haptic info: ");
+		return FALSE;
+	}
+	if (!fu_elantp_hid_haptic_device_read_cmd(parent,
+						  ETP_CMD_I2C_EEPROM_SETTING,
+						  buf,
+						  sizeof(buf),
+						  error)) {
+		g_prefix_error(error, "failed to read haptic info: ");
+		return FALSE;
+	}
+	value = fu_memread_uint16(buf, G_LITTLE_ENDIAN);
+
+	if ((value & 0xFFFF) != ETP_CMD_I2C_EEPROM_WRITE_INFOMATION) {
+		g_prefix_error(error, "failed to set haptic info (0x%04x): ", value);
+		return FALSE;
+	}
+	if (!fu_elantp_hid_haptic_device_write_cmd(parent,
+						   ETP_CMD_I2C_IAP,
+						   helper->iap_password,
+						   error)) {
+		g_prefix_error(error, "failed to write iap password: ");
+		return FALSE;
+	}
+	if (!fu_elantp_hid_haptic_device_write_cmd(parent,
+						   ETP_CMD_I2C_EEPROM_WRITE_CHECKSUM,
+						   helper->checksum,
+						   error)) {
+		g_prefix_error(error, "failed to write eeprom checksum: ");
+		return FALSE;
+	}
+	if (!fu_elantp_hid_haptic_device_write_cmd(parent,
+						   ETP_CMD_I2C_EEPROM_SETTING,
+						   ETP_CMD_I2C_EEPROM_SETTING_INITIAL,
+						   error)) {
+		g_prefix_error(error, "failed to set haptic initial setting: ");
+		return FALSE;
+	}
+	if (!fu_elantp_hid_haptic_device_read_cmd(parent,
+						  ETP_CMD_I2C_EEPROM_WRITE_CHECKSUM,
+						  buf,
+						  sizeof(buf),
+						  error)) {
+		g_prefix_error(error, "failed to read haptic checksum: ");
+		return FALSE;
+	}
+	value = fu_memread_uint16(buf, G_LITTLE_ENDIAN);
+
+	if ((value & 0xFFFF) != helper->checksum) {
+		g_prefix_error(error,
+			       "eeprom checksum failed 0x%04x != 0x%04x : ",
+			       value,
+			       helper->checksum);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+static gboolean
+fu_elantp_hid_haptic_device_get_checksum(FuDevice *parent, guint16 *checksum, GError **error)
+{
+	guint16 ctrl = 0;
+	int cnt = 0;
+	guint8 buf[2] = {0x0};
+
+	if (!fu_elantp_hid_haptic_device_write_cmd(parent,
+						   ETP_CMD_I2C_SET_EEPROM_CTRL,
+						   ETP_CMD_I2C_CALC_EEPROM_CHECKSUM,
+						   error))
+		return FALSE;
+
+wait:
+	g_usleep(ELANTP_EEPROM_READ_DELAY * 1000);
+	if (!fu_elantp_hid_haptic_device_write_cmd(parent,
+						   ETP_CMD_I2C_SET_EEPROM_CTRL,
+						   ETP_CMD_I2C_SET_EEPROM_DATATYPE,
+						   error))
+		return FALSE;
+
+	if (!fu_elantp_hid_haptic_device_read_cmd(parent,
+						  ETP_CMD_I2C_SET_EEPROM_CTRL,
+						  buf,
+						  sizeof(buf),
+						  error)) {
+		g_prefix_error(error, "failed to read calc haptic cmd: ");
+		return FALSE;
+	}
+	ctrl = fu_memread_uint16(buf, G_LITTLE_ENDIAN);
+
+	if ((ctrl & 0x20) == 0x20) {
+		cnt++;
+		if (cnt >= 100) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_READ,
+				    "failed to wait calc haptic cmd");
+			return FALSE;
+		}
+		goto wait;
```

Comment:
@jinglewu I don't like this; can we use `fu_device_retry` here as well please.

Reply:
I used the fu_device_retry_full() or fu_device_retry() function, I set it 100 times and it only runs once.